packages/@lwc/engine-core/src/framework/membrane.ts

```diff
@@ -21,5 +21,24 @@ export const reactiveMembrane = new ObservableMembrane({
  * change or being removed.
  */
 export function unwrap(value: any): any {
-    return reactiveMembrane.unwrapProxy(value);
+    // On the server side, we don't need mutation tracking. Skipping it improves performance.
+    return process.env.IS_BROWSER ? reactiveMembrane.unwrapProxy(value) : value;
+}
+
+export function getReadOnlyProxy(value: any): any {
+    // On the server side, we don't need mutation tracking. Skipping it improves performance.
+    return process.env.IS_BROWSER ? reactiveMembrane.getReadOnlyProxy(value) : value;
```

Comment:
this is the only piece that concerns me. The fact that in SSR the value passed down to you can be mutated it is a problem.

Reply:
Hm, should we return a frozen copy instead?

Reply:
Hm, well I tried to write my own lightweight deep-frozen proxy, but it's not as fast as `observable-membrane`. 😅 So I guess for `getReadOnlyProxy`, we still need to defer to `observable-membrane` on both the client and the server.

The performance is still improved by 9%-12%, which is not as good as 13%-18%, but it's still better than 0.

![Screen Shot 2022-07-29 at 3 17 08 PM](https://user-images.githubusercontent.com/283842/181852743-8f2f7551-4b87-4511-9538-333df4b90cc7.png)
![Screen Shot 2022-07-29 at 3 17 13 PM](https://user-images.githubusercontent.com/283842/181852744-45b9413c-caca-445f-bc33-2fe721da7fe3.png)

I also added a test to confirm that children cannot mutate props passed in by their parents.

Reply:
@caridy What do you think if maybe the server components could run in a kind of "validation mode" that would include the read-only proxy checks, but in "prod mode", the read-only proxy would be disabled?

There are competing desires here: we _should_ disallow mutations on the properties, especially because it matches the client-side behavior, but if you want your SSR to perform as well as possible, and if all your components are well-behaved, then it's just extra overhead. ~6% can be a big deal in terms of server throughput.

I suppose we could make the same argument for the client-side, e.g. only doing the read-only proxy for `process.env.NODE_ENV === 'development'`.

Reply:
Na', I think we should stay away from that kind of non-deterministic behavior mechanism based on the environment.